### ISSUE TITLE
unhandled 'error' event crashes maildev shortly after viewing email in gui

### ISSUE THREAD

**Issue by heldchen** [maildev/maildev#287]
there's an unhandled `ETIMEDOUT` error that seems to crash maildev shortly (about a minute?) after received emails are read in the webgui.:

```
MailDev outgoing SMTP Server my-real-server.com:25 (user:undefined, pass:undefined, secure:no)
Auto-Relay mode on, Relay rules: [{"deny":"*"},{"allow":"my@mail.com"}]
MailDev webapp running at http://:::1080
MailDev SMTP Server running at :::1025
events.js:288
      throw er; // Unhandled 'error' event
      ^

Error: Connection timeout
    at SMTPConnection._formatError (/usr/lib/node_modules/maildev/node_modules/nodemailer/lib/smtp-connection/index.js:557:19)
    at SMTPConnection._onError (/usr/lib/node_modules/maildev/node_modules/nodemailer/lib/smtp-connection/index.js:530:20)
    at Timeout.<anonymous> (/usr/lib/node_modules/maildev/node_modules/nodemailer/lib/smtp-connection/index.js:248:18)
    at listOnTimeout (internal/timers.js:549:17)
    at processTimers (internal/timers.js:492:7)
Emitted 'error' event on SMTPConnection instance at:
    at SMTPConnection._onError (/usr/lib/node_modules/maildev/node_modules/nodemailer/lib/smtp-connection/index.js:547:14)
    at Timeout.<anonymous> (/usr/lib/node_modules/maildev/node_modules/nodemailer/lib/smtp-connection/index.js:248:18)
    at listOnTimeout (internal/timers.js:549:17)
    at processTimers (internal/timers.js:492:7) {
  code: 'ETIMEDOUT',
  command: 'CONN'
}
MailDev outgoing SMTP Server my-real-server.com:25 (user:undefined, pass:undefined, secure:no)
Auto-Relay mode on, Relay rules: [{"deny":"*"},{"allow":"my@mail.com"}]
MailDev webapp running at http://:::1080
MailDev SMTP Server running at :::1025
events.js:288
      throw er; // Unhandled 'error' event
      ^

Error: Connection timeout
    at SMTPConnection._formatError (/usr/lib/node_modules/maildev/node_modules/nodemailer/lib/smtp-connection/index.js:557:19)
    at SMTPConnection._onError (/usr/lib/node_modules/maildev/node_modules/nodemailer/lib/smtp-connection/index.js:530:20)
    at Timeout.<anonymous> (/usr/lib/node_modules/maildev/node_modules/nodemailer/lib/smtp-connection/index.js:248:18)
    at listOnTimeout (internal/timers.js:549:17)
    at processTimers (internal/timers.js:492:7)
Emitted 'error' event on SMTPConnection instance at:
    at SMTPConnection._onError (/usr/lib/node_modules/maildev/node_modules/nodemailer/lib/smtp-connection/index.js:547:14)
    at Timeout.<anonymous> (/usr/lib/node_modules/maildev/node_modules/nodemailer/lib/smtp-connection/index.js:248:18)
    at listOnTimeout (internal/timers.js:549:17)
    at processTimers (internal/timers.js:492:7) {
  code: 'ETIMEDOUT',
  command: 'CONN'
}
MailDev outgoing SMTP Server my-real-server.com:25 (user:undefined, pass:undefined, secure:no)
Auto-Relay mode on, Relay rules: [{"deny":"*"},{"allow":"my@mail.com"}]
MailDev webapp running at http://:::1080
MailDev SMTP Server running at :::1025
```
the crash seems to happen around 60s after viewing an email in the gui.

version 1.1.0 installed through npm on a ubuntu 19.10 system.

**Comment by djfarrelly** [comment]
@heldchen, do you think this could be related to #252 and it's fix in  #275?

2.0.0-beta2 is out now and it has the fix. Would you be up for giving it a try?

```
npm i maildev@2.0.0-beta2
```

**Comment by heldchen** [comment]
@djfarrelly thanks, I'll give it a try, but from the first quick look the underlying error seems to be a different one: I've in the meantime traced it down to SMTPConnection not handling protocol "errors" (in quotes, because most of the failures seem to be from unknown commands/responses from remote servers) at all, instead it emits an fatal 'error' event.

I've worked around this by changing `lib/outgoing.js` to set an onError function. adding 
```
client.on('error', function(e) { logger.error('Caught SMTP protocol error: ', e) })
```
to outgoing.js's `_createClient()` made maildev stable for me (i.e. no service crashes for 30+ days):
```
outgoing._createClient = function () {
  try {
    client = new SMTPConnection({
      port: config.port,
      host: config.host,
      secure: config.secure,
      auth: (config.pass && config.user) ? { user: config.user, pass: config.pass } : false,
      tls: { rejectUnauthorized: false },
      debug: true
    })

    client.on('error', function(e) { logger.error('Caught SMTP protocol error: ', e) })

    logger.info(
      'MailDev outgoing SMTP Server %s:%d (user:%s, pass:%s, secure:%s)',
      config.host,
      config.port,
      config.user,
      config.pass ? '####' : config.pass,
      config.secure ? 'yes' : 'no'
    )
  } catch (err) {
    logger.error('Error during configuration of SMTP Server for outgoing email', err)
  }
}
```

**Comment by djfarrelly** [comment]
@heldchen, thanks for sharing this - this really helps! The docs ([link](https://nodemailer.com/extras/smtp-connection/#events)) say that the connection is automatically closed during an error, but the app calls `connect()` when each email is relayed in `relayMail()`:

https://github.com/maildev/maildev/blob/6d0b5e2550099b995bede5c9f5b955d397c03e08/lib/outgoing.js#L201

I think that should allow the app to reconnect after experiencing errors. Is that what you experience? 

If this is the case, I'll take your change and create a PR and release another 2.0.0 beta 😄

**Comment by heldchen** [comment]
> I think that should allow the app to reconnect after experiencing errors. Is that what you experience?

yes, indeed - reconnecting is not an issue. with my workaround in place, I have not had any issues at all, be it with receiving or sending mails.

I did have the SMTPConnection crashes at several places tho, not just the relay function (for which the stacktrace is in this issue), thus the general onError approach. for example I had the unhandled errors when portscanners were testing connections or when maildev is configure as exim4's smarthost target server. with the onError in place, maildev is working fine in all my scenarios, even relaying. as you pointed out, the connection is automatically reestablished.

I'm not that familiar with node event programming, but as far as I can tell my issue was due to SMTPConnection._onError escalating the error upstream: I've first tried to subscribe to the STMPConnection error event in outgoing.js and then use `done(error)` but haven't had much luck - it seemed as soon as the SMTPConnection._onError finished, the nodejs internal event.js takes over and throws the fatal "unhandled error event" stopping the entire app. maybe you know of a better way to handle this directly in outgoing.js without injecting error handler functions into SMTPConnection?

lastly: thanks for your work on maildev, really useful tool for local development :-)
